backpack/__init__.py

```diff
@@ -1,44 +1,49 @@
 """BackPACK."""
 import inspect
+from typing import Callable, Tuple, Union
 
 import torch
+from torch import Tensor
+from torch.nn import Module
 
 from backpack.extensions.backprop_extension import BackpropExtension
 from backpack.utils.hooks import no_op
 
 from . import extensions
 from .context import CTX
+from .utils import TORCH_VERSION_HIGHER_THAN_1_9_0
+from .utils.module_classification import is_no_op
 
 
 class backpack:
-    """Activate BackPACK extensions.
+    """Activate BackPACK extensions."""
 
-    Enables the BackPACK extensions passed as arguments in the
-    :code:`backward` calls inside the current :code:`with` block.
+    def __init__(
+        self,
+        *exts: BackpropExtension,
+        extension_hook: Callable[[Module], None] = None,
+        debug: bool = False
+    ):
+        """Activate BackPACK extensions.
 
-    Args:
-        exts ([BackpropExtension]): Extensions to activate in the backward pass.
-        extension_hook (function, optional): Function called on each module after
-            all BackPACK extensions have run. Takes a ``torch.nn.Module`` and returns
-            ``None``. Default: ``None`` (no operation will be formed).
+        Enables the BackPACK extensions passed as arguments in the
+        :code:`backward` calls inside the current :code:`with` block.
+
+        Args:
+            exts: Extensions to activate in the backward pass.
+            extension_hook: Function called on each module after
+                all BackPACK extensions have run. Takes a ``torch.nn.Module`` and returns
+                ``None``. Default: ``None`` (no operation will be formed).
+            debug: Print debug messages during the backward pass. Default: ``False``.
 
-            Can be used to reduce memory overhead if the goal is to compute
+        .. note::
+            extension_hook can be used to reduce memory overhead if the goal is to compute
             transformations of BackPACK quantities. Information can be compacted
             during a backward pass and obsolete tensors be freed manually (``del``).
 
-            .. note::
-
-                If the callable iterates over the ``module.parameters()``, the same
-                parameter may be seen multiple times across calls. This happens
-                if the parameters are part of multiple modules.
-                For example, the parameters of a `torch.nn.Linear` module in
-                ``model = torch.nn.Sequential(torch.nn.Linear(...))`` are part of
-                both the ``Linear`` and the ``Sequential``.
```

Comment:
Needs confirmation by a test: Extension hook visits every parameter once and is not called on containers which would lead to visiting the contained parameters more than once.